File: write-api-server-kotlin/src/main/kotlin/com/github/j5ik2o/adceet/api/write/use/case/CreateThreadUseCaseImpl.kt
package com.github.j5ik2o.adceet.api.write.use.case

import akka.actor.typed.ActorRef
import akka.actor.typed.ActorSystem
import akka.actor.typed.javadsl.AskPattern
import com.github.j5ik2o.adceet.api.write.adaptor.aggregate.ThreadAggregateProtocol
import com.github.j5ik2o.adceet.api.write.domain.AccountId
import com.github.j5ik2o.adceet.api.write.domain.ThreadId
import java.time.Duration
import java.util.UUID
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletionStage

class CreateThreadUseCaseImpl(
  private val system: ActorSystem<Void>,
  private val threadAggregateRef: ActorRef<ThreadAggregateProtocol.CommandRequest>,
  private val askTimeout: Duration = Duration.ofSeconds(30)
) : CreateThreadUseCase {
  override fun execute(threadId: ThreadId, accountId: AccountId): CompletionStage<ThreadId> {
    return AskPattern.ask<ThreadAggregateProtocol.CommandRequest, ThreadAggregateProtocol.CreateThreadReply>(
      threadAggregateRef,
      { replyTo -> ThreadAggregateProtocol.CreateThread(UUID.randomUUID(), threadId, accountId, replyTo) },
      askTimeout,
      system.scheduler()
    ).thenCompose { result ->
      when (result) {
        is ThreadAggregateProtocol.CreateThreadSucceeded ->
          CompletableFuture.completedStage(result.threadId)
        is ThreadAggregateProtocol.CreateThreadFailed ->
          CompletableFuture.failedFuture(CreateThreadException(result.error.message))
      }
    }
  }
}

File: write-api-server-scala/src/main/scala/com/github/j5ik2o/adceet/api/write/use/case/CreateThreadUseCaseImpl.scala
package com.github.j5ik2o.adceet.api.write.use.`case`

import akka.actor.typed.scaladsl.AskPattern._
import akka.actor.typed.{ ActorRef, ActorSystem, Scheduler }
import akka.util.Timeout
import com.github.j5ik2o.adceet.api.write.aggregate.ThreadAggregateProtocol
import com.github.j5ik2o.adceet.api.write.domain.{ AccountId, ThreadId }
import wvlet.airframe.ulid.ULID

import scala.concurrent.duration.{ DurationInt, FiniteDuration }
import scala.concurrent.{ ExecutionContext, Future }

final class CreateThreadUseCaseImpl(
    private val system: ActorSystem[_],
    private val threadAggregateRef: ActorRef[ThreadAggregateProtocol.CommandRequest],
    private val askTimeout: FiniteDuration = 120.seconds
) extends CreateThreadUseCase {

  override def execute(threadId: ThreadId, accountId: AccountId)(implicit ec: ExecutionContext): Future[ThreadId] = {
    implicit val to: Timeout          = askTimeout
    implicit val scheduler: Scheduler = system.scheduler
    threadAggregateRef
      .ask(replyTo => ThreadAggregateProtocol.CreateThread(ULID.newULID, threadId, accountId, replyTo)).flatMap {
        case ThreadAggregateProtocol.CreateThreadSucceeded(_, _, threadId) =>
          Future.successful(threadId)
        case ThreadAggregateProtocol.CreateThreadFailed(_, _, _, error) =>
          Future.failed(new CreateThreadException(error.message))
      }
  }

}
